server/src/main/java/org/opensearch/index/codec/Lucene99CoreStoredFieldsFormat.java

```diff
@@ -0,0 +1,147 @@
+        this(Lucene99Codec.Mode.BEST_SPEED);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param mode The mode represents ZSTD or ZSTDNODICT
+     */
+
+    public Lucene99CoreStoredFieldsFormat(Lucene99Codec.Mode mode) {
+        this.mode = Objects.requireNonNull(mode);
+    }
+
+    /**
+     * Returns a {@link StoredFieldsReader} to load stored fields.
+     *
+     * @param directory The index directory.
+     * @param si        The SegmentInfo that stores segment information.
+     * @param fn        The fieldInfos.
+     * @param context   The IOContext that holds additional details on the merge/search context.
+     */
+    @Override
+    public StoredFieldsReader fieldsReader(Directory directory, SegmentInfo si, FieldInfos fn, IOContext context) throws IOException {
+        if (si.getAttribute(MODE_KEY) != null) {
+            String value = si.getAttribute(MODE_KEY);
+            Lucene99Codec.Mode mode = Lucene99Codec.Mode.valueOf(value);
+            return impl(mode).fieldsReader(directory, si, fn, context);
+        } else {
+            throw new IllegalStateException("missing value for " + MODE_KEY + " for segment: " + si.name);
+        }
+
+    }
+
+    /**
+     * Returns a {@link StoredFieldsReader} to write stored fields.
+     *
+     * @param directory The index directory.
+     * @param si        The SegmentInfo that stores segment information.
+     * @param context   The IOContext that holds additional details on the merge/search context.
+     */
+
+    @Override
+    public StoredFieldsWriter fieldsWriter(Directory directory, SegmentInfo si, IOContext context) throws IOException {
+        String previous = si.putAttribute(MODE_KEY, mode.name());
+        if (previous != null && previous.equals(mode.name()) == false) {
+            throw new IllegalStateException(
+                "found existing value for " + MODE_KEY + " for segment: " + si.name + "old=" + previous + ", new=" + mode.name()
+            );
+        }
+        return impl(mode).fieldsWriter(directory, si, context);
+    }
+
+    StoredFieldsFormat impl(Lucene99Codec.Mode mode) {
+        switch (mode) {
+            case BEST_SPEED:
+                return getLZ4CompressingStoredFieldsFormat();
+            case BEST_COMPRESSION:
+                return getZlibCompressingStoredFieldsFormat();
+            default:
+                throw new AssertionError();
+        }
+    }
+
+    public Lucene99Codec.Mode getMode() {
+        return mode;
+    }
+
+    // Shoot for 10 sub blocks of 48kB each.
+    private static final int BEST_COMPRESSION_BLOCK_LENGTH = 10 * 48 * 1024;
+
+    /**
+     * Compression mode for {@link Lucene90StoredFieldsFormat.Mode#BEST_COMPRESSION}
+     */
+    public static final CompressionMode BEST_COMPRESSION_MODE = new DeflateWithPresetDictCompressionMode();
+
+    // Shoot for 10 sub blocks of 8kB each.
+    private static final int BEST_SPEED_BLOCK_LENGTH = 10 * 16 * 1024;
```

Comment:
The comment says 8kB while the value is 16kB.

What is the effect of this change w.r.t i) disk size, ii) indexing performance and iii) search performance?

Reply:
Also, what is the ongoing maintenance burden of this change? It looks like we'll be on the hook to implement new codec wrappers on every Lucene minor version. And is it plausible that future version of Lucene codecs might not expose the configurability we rely on here and then leave us in a bad spot regarding compatibility?

Reply:
thanks folks for taking a look. This PR should have ideally been in draft stage ~ have moved it to draft.
Thanks for bringing these points up, I'm working on the fresh benchmarks, and correctness as of now. 
Will get back on these questions in some time!

Reply:
> It looks like we'll be on the hook to implement new codec wrappers on every Lucene minor version.

Yes, we will be maintaining new codec wrappers. Let me think if there is another way to get around this.

Reply:
Also, should we let customers control the page size? For example, MySQL lets customers configure the page size (https://dev.mysql.com/doc/refman/8.0/en/innodb-parameters.html#sysvar_innodb_page_size)